File: packages/sdk-python/src/acruxcore/response_format.py
"""Build a ``response_format`` from a pydantic model — the typed alternative to the dict.

The SDK's :data:`~acruxcore.types.ResponseFormat` is the OpenAI-shaped dict the gateway
forwards straight to the provider. This module offers a typed way to produce that same
dict from a pydantic v2 ``BaseModel`` subclass, so field-level ``Field(description=...)``
guidance reaches the model without hand-writing JSON Schema. Pydantic is an *optional*
dependency — it is imported lazily only when a pydantic-built format is actually sent, and
:func:`pydantic_response_format` itself does not import it (the import happens at send time
in :func:`normalize_response_format`).
"""

from __future__ import annotations

from typing import Any, Dict, Optional, Type

from .errors import PYDANTIC_NOT_AVAILABLE, AcruxCoreError

#: Sentinel key marking a dict as "build me from a pydantic model at send time". A private
#: contract between :func:`pydantic_response_format` and :func:`normalize_response_format`;
#: callers build these only via the helper.
_PYDANTIC_SENTINEL = "__acruxcore_pydantic_response_format__"


def pydantic_response_format(
    model: Type[Any],
    *,
    name: str,
    strict: bool = True,
) -> Dict[str, Any]:
    """Build a ``response_format`` from a pydantic v2 ``BaseModel`` subclass.

    Returns a marker dict the SDK resolves to the OpenAI-shaped wire format when the
    request is sent (calling ``model.model_json_schema()``). Pydantic itself is imported
    only at that point, so this call never requires pydantic to be installed.

    :param model: A pydantic v2 ``BaseModel`` subclass (the class itself, not an instance).
    :param name: The OpenAI ``json_schema.name`` for the request — required, since a class
        name is an API contract on the gateway, not a label the SDK should guess. Lowercase
        snake_case is conventional (e.g. ``"medical_information_answer"``).
    :param strict: Whether to ask the provider for strict schema adherence (default
        ``True``). Maps to ``json_schema.strict``.
    :returns: An opaque marker dict; pass it as ``response_format=`` to
        :meth:`~acruxcore.client.AcruxCore.chat` or
        :meth:`~acruxcore.client.AcruxCore.run_tool_loop`.
    """
    if not isinstance(model, type):
        raise AcruxCoreError(
            "acruxcore: pydantic_response_format() takes a BaseModel subclass (the class "
            f"itself), received an instance of {type(model).__name__}.",
            "TOOL_SCHEMA_ERROR",
        )
    if not name or not isinstance(name, str):
        raise AcruxCoreError(
            "acruxcore: pydantic_response_format() requires a non-empty `name` for the "
            "gateway's json_schema.name field.",
            "TOOL_SCHEMA_ERROR",
        )
    return {_PYDANTIC_SENTINEL: model, "name": name, "strict": strict}


def is_pydantic_response_format(value: Any) -> bool:
    """True when ``value`` is a marker dict from :func:`pydantic_response_format`."""
    return (
        isinstance(value, dict)
        and _PYDANTIC_SENTINEL in value
        and isinstance(value[_PYDANTIC_SENTINEL], type)
    )


def normalize_response_format(
    response_format: Optional[Dict[str, Any]],
) -> Optional[Dict[str, Any]]:
    """Resolve a ``response_format`` to the OpenAI-shaped wire dict.

    A pydantic-built marker dict (from :func:`pydantic_response_format`) is converted by
    importing pydantic lazily and calling ``model_json_schema()``; anything else is passed
    through unchanged. Returns ``None`` when ``response_format`` is ``None``.

    :raises AcruxCoreError: ``PYDANTIC_NOT_AVAILABLE`` if a pydantic-built format was given
        but pydantic is not installed.
    """
    if response_format is None:
        return None
    if not is_pydantic_response_format(response_format):
        return response_format

    model: Type[Any] = response_format[_PYDANTIC_SENTINEL]
    try:
        schema = model.model_json_schema()
    except AttributeError:
        raise AcruxCoreError(
            "acruxcore: the class passed to pydantic_response_format() has no "
            "model_json_schema() — it is not a pydantic v2 BaseModel subclass. "
            f"Got {model.__name__ if isinstance(model, type) else type(model).__name__}.",
            "TOOL_SCHEMA_ERROR",
        )
    except ImportError as exc:
        raise AcruxCoreError(
            "acruxcore: a pydantic-built response_format was given but pydantic could not "
            f"be imported ({exc}). Install pydantic (>=2) alongside the SDK, or pass a "
            "plain JSON Schema dict as response_format instead.",
            PYDANTIC_NOT_AVAILABLE,
        )

    strict = response_format.get("strict", True)
    # pydantic's model_json_schema() does not emit the keys OpenAI's strict mode requires
    # (`additionalProperties: false` on every object) and emits `title` keys it rejects.
    # Normalize when strict is set so the default actually works against OpenAI; leave the
    # schema untouched when the caller opted out of strict (Anthropic/Gemini are lenient).
    if strict:
        schema = _strict_for_openai(schema)

    return {
        "type": "json_schema",
        "json_schema": {
            "name": response_format["name"],
            "schema": schema,
            "strict": strict,
        },
    }


def _strict_for_openai(node: Any) -> Any:
    """Make a pydantic JSON Schema acceptable to OpenAI's strict `response_format`.

    OpenAI rejects a `strict: true` schema unless every object sets
    ``additionalProperties: false``; it also rejects the ``title`` keys pydantic adds to
    every field and the top level. A pydantic ``BaseModel`` is always a closed object, so
    adding ``additionalProperties: false`` is faithful, not a behavior change. Walks the
    schema recursively (objects, arrays, ``$defs``, ``allOf``/``anyOf``/``oneOf``).
    """
    if isinstance(node, list):
        return [_strict_for_openai(n) for n in node]
    if not isinstance(node, dict):
        return node
    cleaned = {k: _strict_for_openai(v) for k, v in node.items() if k != "title"}
    # `title` is the only key OpenAI strict mode rejects that pydantic always adds; keys
    # like `description` / `default` / `$defs` are fine and left in place.
    if cleaned.get("type") == "object" and "additionalProperties" not in cleaned:
        cleaned["additionalProperties"] = False
    return cleaned

File: packages/sdk/src/responseFormat.ts
import { acruxcoreError } from './error';
import { isZodV4Schema, loadZodToJsonSchema } from './tools';
import type { ResponseFormat } from './types';

/**
 * Resolves a {@link ResponseFormat} into the OpenAI-shaped wire dict the gateway forwards
 * to the provider. The `{ zod, name }` form is converted here by importing zod dynamically
 * (it is an optional peer dependency) and calling `toJSONSchema`; every other form passes
 * straight through unchanged.
 *
 * Mirrors `resolveParametersSchema` in `tools.ts` — same dynamic-import discipline, same
 * zod v3-vs-v4 guard, same error codes. Run this ONCE at the public entrypoint of `chat()`
 * / `runToolLoop()` so the rest of the pipeline (the synchronous `_buildChatBody`) only
 * ever sees the plain wire dict.
 *
 * @param responseFormat - The caller's `ResponseFormat`, or `undefined`.
 * @returns The resolved OpenAI-shaped dict, or `undefined` when none was given.
 * @throws {acruxcoreError} `ZOD_NOT_AVAILABLE` if a zod schema was given but zod cannot be
 *   imported; `TOOL_SCHEMA_ERROR` if it is a classic zod v3 schema.
 */
export async function resolveResponseFormat(
  responseFormat: ResponseFormat | undefined,
): Promise<Record<string, unknown> | undefined> {
  if (!responseFormat) return undefined;
  // The `{ zod, name }` arm is the only one that needs conversion.
  if ('zod' in responseFormat) {
    if (!isZodV4Schema(responseFormat.zod)) {
      throw new acruxcoreError(
        'acruxcore: this looks like a classic zod v3 schema, which cannot be converted to JSON ' +
          "Schema. Import zod's v4 API instead — `import { z } from 'zod/v4'` on zod 3.25+, or " +
          "`import { z } from 'zod'` on zod 4 — or pass a plain { type: 'json_schema', ... } " +
          'object as `responseFormat` instead.',
        'TOOL_SCHEMA_ERROR',
      );
    }
    const toJSONSchema = await loadZodToJsonSchema();
    const schema = toJSONSchema(responseFormat.zod) as Record<string, unknown>;
    const { $schema: _ignored, ...rest } = schema;
    return {
      type: 'json_schema',
      json_schema: { name: responseFormat.name, schema: rest, strict: responseFormat.strict ?? true },
    };
  }
  // The text / json_object / json_schema forms are already the wire shape.
  return responseFormat as Record<string, unknown>;
}
